tdengine/src/tdengine.rs

```diff
@@ -0,0 +1,1058 @@
+    /// #Returns
+    /// An emtpy result or an error arrising a malformed ``DigitizerAnalogTraceMessage`` parameter.
+    async fn process_message(
+        &mut self,
+        message: &DigitizerAnalogTraceMessage,
+    ) -> Result<(), error::Error> {
+        // Obtain the channel data, and error check
+        self.error.test_metadata(message);
+
+        // Obtain message data, and error check
+        self.frame_data.init(message)?;
+
+        // Obtain the channel data, and error check
+        self.error
+            .test_channels(&self.frame_data, &message.channels().unwrap());
+
+        let table_name = self.frame_data.get_table_name();
+        let frame_table_name = self.frame_data.get_frame_table_name();
+        let frame_column_views = views::create_frame_column_views(
+            &self.frame_data,
+            &self.error,
+            &message.channels().unwrap(),
+        );
+        let column_views =
+            views::create_column_views(&self.frame_data, &message.channels().unwrap());
+        let tags = [Value::UTinyInt(self.frame_data.digitizer_id)];
+
+        //  Initialise Statement
+        self.stmt
+            .set_tbname(table_name)
+            .map_err(|e| TDEngineError::Stmt(StatementError::SetTableName, e))?;
+        self.stmt
+            .set_tags(&tags)
+            .map_err(|e| TDEngineError::Stmt(StatementError::SetTags, e))?;
+        self.stmt
+            .bind(&column_views)
+            .map_err(|e| TDEngineError::Stmt(StatementError::Bind, e))?;
+        self.stmt
+            .add_batch()
+            .map_err(|e| TDEngineError::Stmt(StatementError::AddBatch, e))?;
+
+        self.frame_stmt
+            .set_tbname(frame_table_name)
+            .map_err(|e| TDEngineError::Stmt(StatementError::SetTableName, e))?;
+        self.frame_stmt
+            .set_tags(&tags)
+            .map_err(|e| TDEngineError::Stmt(StatementError::SetTags, e))?;
+        self.frame_stmt
+            .bind(&frame_column_views)
+            .map_err(|e| TDEngineError::Stmt(StatementError::Bind, e))?;
+        self.frame_stmt
+            .add_batch()
+            .map_err(|e| TDEngineError::Stmt(StatementError::AddBatch, e))?;
+        Ok(())
+    }
+
+    /// Sends data extracted from a previous call to ``process_message`` to the tdengine server.
+    /// #Returns
+    /// The number of rows affected by the post or an error
+    async fn post_message(&mut self) -> Result<usize, error::Error> {
+        self.stmt
+            .execute()
+            .map_err(|e| error::Error::TDEngine(TDEngineError::Stmt(StatementError::Execute, e)))
+            .unwrap();
+        self.frame_stmt
+            .execute()
+            .map_err(|e| error::Error::TDEngine(TDEngineError::Stmt(StatementError::Execute, e)))
+            .unwrap();
+        Ok(0)
+    }
+}
+
+// Note these tests should be run in single-threaded mode, in an environment with a .env file in the working directory and a TDEngine server on the network.
```

Comment:
This can be left for later, but since this file is more test than code, you can consider splitting the tests into their own files.

(e.g. https://github.com/DanNixon/satori/blob/9d4ab2bdcc02911bb249b76d5fe40d707c624578/storage/src/providers/mod.rs#L5, not exactly the same thing, but the best example I had)